File: software/ios/Tucklet/Connectivity/SessionManager.swift
// SessionManager.swift
// Bridges the control plane (BLE) to the data plane (Wi-Fi). Resolves which
// transport to use for this device + platform, brings the network up, and
// hands back a ready-to-use DataClient.
//
// License: PolyForm Noncommercial 1.0.0

import Foundation
import NetworkExtension

public enum SessionError: Error { case noUsableTransport, joinFailed(String), wiredNotOnIOS }

@MainActor
public final class SessionManager {
    private let ble: BLEControlClient
    public init(ble: BLEControlClient) { self.ble = ble }

    /// Mirror of `tucklet-core::variant::usable_transports` for the iOS client.
    public func usableTransports(_ caps: DeviceCapabilities) -> [DataTransport] {
        var out: [DataTransport] = []
        if caps.supports(.wifiAware) { out.append(.wifiAware) }   // iOS 26+ supports Aware
        if caps.supports(.softAp) { out.append(.softAp) }
        // Wired (USB-C bridge) appears as a mass-storage device handled by the
        // Files app / File Provider, not through this Wi-Fi session manager.
        return out
    }

    /// Open a data session: pick transport, get one-time creds over BLE, join.
    public func openDataSession(challengeSignature: String) async throws -> DataClient {
        guard let caps = ble.capabilities else { throw SessionError.noUsableTransport }
        guard let transport = usableTransports(caps).first else { throw SessionError.noUsableTransport }

        let grant = try await ble.openSession(
            SessionRequest(challengeSignature: challengeSignature, transport: transport))

        switch transport {
        case .softAp:
            try await joinSoftAP(ssid: grant.ssidOrService, psk: grant.psk)
        case .wifiAware:
            try await joinWifiAware(service: grant.ssidOrService, passphrase: grant.psk)
        case .wiredUsbHs:
            throw SessionError.wiredNotOnIOS
        }

        let base = URL(string: "http://\(grant.ip)")!
        return DataClient(baseURL: base, token: grant.token)
    }

    /// Join the device's one-time SoftAP. This shows the OS "join Wi-Fi?" prompt
    /// the first time; subsequent joins of the same SSID are smoother.
    private func joinSoftAP(ssid: String, psk: String) async throws {
        let cfg = NEHotspotConfiguration(ssid: ssid, passphrase: psk, isWEP: false)
        cfg.joinOnce = true   // single-use network; don't persist
        do {
            try await NEHotspotConfigurationManager.shared.apply(cfg)
        } catch {
            // "already associated" is success, not failure.
            let ns = error as NSError
            if ns.domain == NEHotspotConfigurationErrorDomain,
               ns.code == NEHotspotConfigurationError.alreadyAssociated.rawValue { return }
            throw SessionError.joinFailed(error.localizedDescription)
        }
    }

    /// Wi-Fi Aware peer-to-peer (iOS 26). Seamless: no "join network" prompt.
    /// CONFIRM the exact WiFiAware framework API (publisher/subscriber, paired
    /// device handle, data path establishment) against the iOS 26 SDK — the
    /// shape below reflects the WWDC25 model but the symbol names must be
    /// verified before this compiles.
    private func joinWifiAware(service: String, passphrase: String) async throws {
        // import WiFiAware
        // let device = try await WAPairedDevice.resolve(...)            // CONFIRM
        // let session = try await WADataSession.connect(to: device,     // CONFIRM
        //                                               service: service,
        //                                               passphrase: passphrase)
        // The DataClient then targets the link-local endpoint the session vends.
        throw SessionError.joinFailed("WiFiAware path pending SDK confirmation")
    }
}

File: software/android/app/src/main/java/app/tucklet/net/SessionManager.kt
// SessionManager.kt
// Orchestrates opening a data session: request it over BLE (with the signed
// challenge), join the charm's Wi-Fi, and hand back a DataClient pinned to it.
// Mirrors tucklet-core::variant transport ordering.
//
// License: PolyForm Noncommercial 1.0.0
package app.tucklet.net

import app.tucklet.ble.BleControlClient
import app.tucklet.protocol.*
import app.tucklet.wifi.WifiBinding
import app.tucklet.wifi.WifiConnector

class SessionManager(
    private val ble: BleControlClient,
    private val wifi: WifiConnector,
) {
    private var binding: WifiBinding? = null

    /** Conservative transport ordering: Aware first (seamless), then SoftAP. We
     *  never auto-pick the wired path for a wireless session. */
    fun usableTransports(caps: DeviceCapabilities): List<DataTransport> {
        val order = listOf(DataTransport.WIFI_AWARE, DataTransport.SOFT_AP)
        return order.filter { caps.supports(it) }
    }

    /**
     * Open a data session and return a connected DataClient.
     * On Android both Aware and SoftAP resolve to a WifiNetworkSpecifier join
     * (the grant carries the SSID/service + PSK either way).
     */
    suspend fun openDataSession(caps: DeviceCapabilities, challengeSignature: String): DataClient {
        val transport = usableTransports(caps).firstOrNull() ?: DataTransport.SOFT_AP
        val grant: SessionGrant = ble.openSession(SessionRequest(challengeSignature, transport))
        val b = wifi.join(grant.ssidOrService, grant.psk)
        binding = b
        return DataClient(b.network, grant.ip, grant.token)
    }

    /** End the session: release the Wi-Fi binding (the charm tears down the AP
     *  on its side when the session times out / disconnects). */
    fun close() {
        binding?.release(); binding = null
    }
}
